app/services/wallets/update_service.rb

```diff
@@ -61,7 +63,49 @@ def valid_expiration_at?(expiration_at:)
     end
 
     def valid_limitations?
+      result.billable_metrics = billable_metrics
+      result.billable_metric_identifiers = billable_metric_identifiers
       Wallets::ValidateLimitationsService.new(result, **params).valid?
     end
+
+    def process_billable_metrics
+      existing_wallet_billable_metric_ids = wallet.wallet_targets.pluck(:billable_metric_id).compact
+
+      billable_metrics.each do |bm|
+        next if existing_wallet_billable_metric_ids.include?(bm.id)
+
+        WalletTarget.create!(wallet:, billable_metric: bm, organization_id: wallet.organization_id)
+      end
+
+      sanitize_wallet_billable_metrics(existing_wallet_billable_metric_ids) if existing_wallet_billable_metric_ids.present?
+    end
+
+    def sanitize_wallet_billable_metrics(existing_wallet_billable_metric_ids)
+      not_needed_wallet_target_ids = existing_wallet_billable_metric_ids - billable_metrics.pluck(:id)
+      not_needed_wallet_target_ids.each do |wallet_billable_metric_id|
+        WalletTarget.find_by(wallet:, billable_metric_id: wallet_billable_metric_id, organization: wallet.organization)&.destroy!
```

Comment:
I believe wallet target should always be present here since you retrieve here before:

```suggestion
        WalletTarget.find_by(wallet:, billable_metric_id: wallet_billable_metric_id, organization: wallet.organization).destroy!
```

Reply:
You are right, but since query is not performed in the method itself but the method just relies in the variable, I have a practice to leave it. But shouldn't be too expensive :)